bundles/org.palladiosimulator.simulizar/src/org/palladiosimulator/simulizar/modelobserver/AllocationLookupSyncer.java

```diff
@@ -0,0 +1,245 @@
+            simulatedContainerStorage.remove(ctx.getId());
+        } else {
+            var newHierarchy = new LinkedList<AssemblyContext>(ctxHierarchy);
+            newHierarchy.push(ctx);
+            simulatedContainerStorage.remove(new FQComponentID(newHierarchy).getFQIDString());
+            hierarchy = newHierarchy;
+        }
+
+        if (ctx.getEncapsulatedComponent__AssemblyContext() instanceof CompositeComponent) {
+            var composite = (CompositeComponent) ctx;
+            for (var compCtx : composite.getAssemblyContexts__ComposedStructure()) {
+                removeAssemblyAllocation(compCtx, hierarchy);
+            }
+        }
+    }
+
+    /**
+     * This method delegates the processing of the AllocationContexts contained in
+     * the notification to the supplied processor. If the notification is of _MANY
+     * type the processor is called for every contained AllocationContext.
+     * 
+     * @param notification the EMF notification
+     * @param extractor    a function which gets the changed object from the
+     *                     notification.
+     * @param processor    the consumer which processes a single AllocationContext
+     *                     instance
+     */
+    @SuppressWarnings("unchecked")
+    private void processNotification(Notification notification, Function<Notification, Object> extractor,
+            Consumer<AllocationContext> processor) {
+        assert notification.getFeature() == AllocationPackage.Literals.ALLOCATION__ALLOCATION_CONTEXTS_ALLOCATION;
+
+        Object feature = notification.getNewValue();
+        if (feature instanceof Collection) {
+            ((Collection<AllocationContext>) feature).forEach(processor);
+        } else {
+            /*
+             * If the new element is not an allocation context something about the model
+             * changed. Then we will need to revisit this class again anyhow.
+             */
+            doAddAllocationContext((AllocationContext) feature);
+        }
+    }
+
+    /**
+     * Convenience method to add the provided allocation context.
+     */
+    private void doAddAllocationContext(AllocationContext ctx) {
+        addAssemblyAllocation(ctx.getAssemblyContext_AllocationContext(), Collections.emptyList(),
+                resourceContainerAccess.getSimulatedEntity(ctx.getResourceContainer_AllocationContext()));
+    }
+
+    /**
+     * Convenience method to remove the provided allocation context.
+     */
+    private void doRemoveAllocationContext(AllocationContext ctx) {
+        removeAssemblyAllocation(ctx.getAssemblyContext_AllocationContext(), Collections.emptyList());
+    }
+
+    @Override
+    protected void add(Notification notification) {
+        if (notification.getFeature() == AllocationPackage.Literals.ALLOCATION__ALLOCATION_CONTEXTS_ALLOCATION) {
+            processNotification(notification, Notification::getNewValue, this::doAddAllocationContext);
+        }
```

Comment:
There is some code duplication in this and the following methods that could be reduced by providing two new methods (one for new value and adding, as well as one for old value and removing) and calling them.

Reply:
I love reducing code duplication